kotlin/samples/dungeon/common/src/main/java/com/squareup/sample/dungeon/GameWorkflow.kt

```diff
@@ -0,0 +1,244 @@
+ */
+package com.squareup.sample.dungeon
+
+import com.squareup.sample.dungeon.ActorWorkflow.ActorInput
+import com.squareup.sample.dungeon.Direction.DOWN
+import com.squareup.sample.dungeon.Direction.LEFT
+import com.squareup.sample.dungeon.Direction.RIGHT
+import com.squareup.sample.dungeon.Direction.UP
+import com.squareup.sample.dungeon.GameWorkflow.GameRendering
+import com.squareup.sample.dungeon.GameWorkflow.Input
+import com.squareup.sample.dungeon.GameWorkflow.Output
+import com.squareup.sample.dungeon.GameWorkflow.Output.PlayerWasEaten
+import com.squareup.sample.dungeon.GameWorkflow.Output.Vibrate
+import com.squareup.sample.dungeon.GameWorkflow.State
+import com.squareup.sample.dungeon.board.Board
+import com.squareup.sample.dungeon.board.Board.Location
+import com.squareup.workflow.RenderContext
+import com.squareup.workflow.Snapshot
+import com.squareup.workflow.StatefulWorkflow
+import com.squareup.workflow.Worker
+import com.squareup.workflow.WorkflowAction.Companion.enterState
+import com.squareup.workflow.onWorkerOutput
+import com.squareup.workflow.renderChild
+import kotlinx.coroutines.delay
+import kotlin.math.roundToLong
+import kotlin.random.Random
+
+private val PAUSED_TICKER = Worker.createSideEffect(key = "noop") {}
+
+class GameWorkflow(
+  private val playerWorkflow: PlayerWorkflow,
+  private val aiWorkflows: List<ActorWorkflow>,
+  private val random: Random
+) : StatefulWorkflow<Input, State, Output, GameRendering>() {
+
+  /**
+   * @param board Should not change while the game is running.
+   */
+  data class Input(
+    val board: Board,
+    val ticksPerSecond: Int = 15
+  )
+
+  /**
+   * @param finishedSnapshot If non-null, the game is finished and this was the last rendering
+   * before the game finished.
+   */
+  data class State(
+    val game: Game
+  )
+
+  sealed class Output {
+    /**
+     * Emitted by [GameWorkflow] if the controller should be vibrated.
+     */
+    object Vibrate : Output()
+
+    object PlayerWasEaten : Output()
+  }
+
+  data class GameRendering(
+    val board: Board,
+    val onPlayerEvent: ((PlayerWorkflow.Event) -> Unit)?
+  )
+
+  override fun initialState(
+    input: Input,
+    snapshot: Snapshot?
+  ): State {
+    val board = input.board
+    return State(game = Game(
+        playerLocation = random.nextEmptyLocation(board),
+        aiLocations = aiWorkflows.map { random.nextEmptyLocation(board) }
+    ))
+  }
+
+  override fun onInputChanged(
+    old: Input,
+    new: Input,
+    state: State
+  ): State {
+    check(old.board == new.board) { "Expected board to not change during the game." }
+    return state
+  }
+
+  override fun render(
+    input: Input,
+    state: State,
+    context: RenderContext<State, Output>
+  ): GameRendering {
+    val running = !state.game.isPlayerEaten
+    // Stop actors from ticking if the game is paused or finished.
+    val ticker = if (running) createTickerWorker(input.ticksPerSecond) else PAUSED_TICKER
```

Comment:
nit: seems odd to create the ticker worker over and over again. Seems like it could be a private var on the input, s.t. it only changes if the tick rate actually changes. Something like that. (Or maybe that's over engineering even for us.)

Reply:
We already do this anywhere you use `asWorker`. We could do that but then we'd need to store the last `ticksPerSecond` or the worker itself in the state. This is less code.